# Email Spam Detection 
# Author : Nischitha D
# Data Science Intern @Oasis Infobyte

In [1]:
import pandas as pd
import numpy as np

In [10]:
df=pd.read_csv("C:/Users/NISCHITHA/Downloads/Spamdata.csv",encoding='latin')
df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [12]:
df.shape

(5572, 5)

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   v1          5572 non-null   object
 1   v2          5572 non-null   object
 2   Unnamed: 2  50 non-null     object
 3   Unnamed: 3  12 non-null     object
 4   Unnamed: 4  6 non-null      object
dtypes: object(5)
memory usage: 217.8+ KB


In [15]:
df.isnull().sum()

v1               0
v2               0
Unnamed: 2    5522
Unnamed: 3    5560
Unnamed: 4    5566
dtype: int64

In [18]:
df.dropna(axis=1, inplace=True)

In [19]:
df.isnull().sum()

v1    0
v2    0
dtype: int64

In [20]:
df.rename(columns={'v1': 'label', 'v2': 'emails'}, inplace=True)

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   label   5572 non-null   object
 1   emails  5572 non-null   object
dtypes: object(2)
memory usage: 87.2+ KB


In [28]:
import pandas as pd
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

In [25]:
X = df['emails']
y = df['label']

In [26]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [29]:
vectorizer = CountVectorizer()
X_train = vectorizer.fit_transform(X_train)
X_test = vectorizer.transform(X_test)

In [30]:
classifier = MultinomialNB()
classifier.fit(X_train, y_train)

MultinomialNB()

In [31]:
y_pred = classifier.predict(X_test)

In [33]:
accuracy = accuracy_score(y_test, y_pred)
confusion = confusion_matrix(y_test, y_pred)
report = classification_report(y_test, y_pred)

In [34]:
print(f'Accuracy: {accuracy:.2f}')
print('\nConfusion Matrix:')
print(confusion)
print('\nClassification Report:')
print(report)

Accuracy: 0.98

Confusion Matrix:
[[963   2]
 [ 16 134]]

Classification Report:
              precision    recall  f1-score   support

         ham       0.98      1.00      0.99       965
        spam       0.99      0.89      0.94       150

    accuracy                           0.98      1115
   macro avg       0.98      0.95      0.96      1115
weighted avg       0.98      0.98      0.98      1115



 # Model as 98% accuracy

# Confusion Matrix:

# 963 emails were correctly classified as 'ham' (non-spam).
# 2 emails were incorrectly classified as 'spam' when they were actually 'ham'.
# 16 emails were incorrectly classified as 'ham' when they were actually 'spam'.
# 134 emails were correctly classified as 'spam'.
    
# Classification Report:

# Precision: The model has a high precision for both 'ham' and 'spam', indicating that when it predicts an email as 'ham' or 'spam', it is usually correct.

# Recall: The model has a high recall for 'ham' (non-spam) but slightly lower recall for 'spam'. This means it's better at identifying non-spam emails but may miss some spam emails.
